polar-core/src/vm.rs

```diff
@@ -223,6 +227,31 @@ impl Default for PolarVirtualMachine {
     }
 }
 
+#[allow(clippy::ptr_arg)]
+fn query_contains_partial(goals: &Goals) -> bool {
+    if goals.len() == 1 {
```

Comment:
This relies on the VM always being instantiated with a single goal.  If there are other goals that contain partials, or a query that occurs later with a partial, it won't be identified.

What if we instead set this flag in the beginning of the `query` function? (The handler for the Query goal.). That could walk the term in the same way, and wouldn't have the same short comings.

Reply:
It doesn't rely on it, it tests for it. The problem with testing in `query` is that it gets hit for every sub-query, of which there are potentially many. This way we do a walk up front, and we're done.

Reply:
Didn't think of the sub-query point.  Perhaps I should have phrased this differently.

If the VM is created with more than one goal, we never check for partials.  This could result in incorrect behavior if the assumption that we always start the VM with one goal is broken.

Instead, we could iterate all initial goals and walk over all query goals to find partials.

Reply:
Ok, done.

Reply:
47dfd0e27bbb6cd7b3540f84a42ed1106fd112bc